### ISSUE TITLE
Multiple OUs

### ISSUE THREAD

**Issue by amaccuish** [mathiasertl/django-ca#38]
I currently have a situation where I would like to have two OUs in the DN. This is allowed by the specs. Django-ca however only selects the last OU from the command line, dropping the other.

**Comment by mathiasertl** [comment]
Yes, the parsing function returns a dictionary, so it's currently only one value. But I've never seen a certificate with multiple OUs. Can you point me to the part of the spec that allows that (so I can look at this in more detail)?

**Comment by amaccuish** [comment]
I can't find it in the spec but it's easily done with OpenSSL and other clis.

[](https://github.com/cloudflare/cfssl/issues/599)
[](https://security.stackexchange.com/questions/39525/generate-csr-with-2-ou-names)

I'll keep looking for proper documentation

**Comment by amaccuish** [comment]
I mean, I guess it's just a X.500 distinguished name, like in LDAP, so you can, I'm trying to find an example of a public certificate, I've seen one recently.

**Comment by mathiasertl** [comment]
Just to be clear: I'm not opposed to supporting this. I just want to know which attributes in the DN can have multiple values etc.

**Comment by amaccuish** [comment]
Sure :) So 100% OU, you can do it in OpenSSL. AFAICT that's the only one. Multiple CNs should use SANs, and the other's don't make sense to have multiple times (e.g. country).

**Comment by amaccuish** [comment]
Here's an example Verisign Issuer DN: "OU = VeriSign Trust Network,OU = "(c) 1998 VeriSign, Inc. - For authorized use only",OU = Class 3 Public Primary Certification Authority - G2,O = "VeriSign, Inc.",C = US"

**Comment by amaccuish** [comment]
![screenshot von 2018-05-26 12 37 20](https://user-images.githubusercontent.com/9434795/40574794-7c5e61e2-60c8-11e8-8fb7-62c38a3156d3.png)

**Comment by mathiasertl** [comment]
Where did you find that cert?

**Comment by amaccuish** [comment]
Here: https://www.tbs-certificates.co.uk/FAQ/en/491.html

Also there's this one:
https://knowledge.symantec.com/support/mpki-for-ssl-support/index?page=content&id=SO5624&actp=LIST&viewlocale=en_US

**Comment by amaccuish** [comment]
Is there anything else you need? :)

**Comment by amaccuish** [comment]
Here's a cloudflare cert:

```
-----BEGIN CERTIFICATE-----
MIIJujCCCWCgAwIBAgIRAJJSmr2F8Kak1sU/0ckQEcEwCgYIKoZIzj0EAwIwgZIx
CzAJBgNVBAYTAkdCMRswGQYDVQQIExJHcmVhdGVyIE1hbmNoZXN0ZXIxEDAOBgNV
BAcTB1NhbGZvcmQxGjAYBgNVBAoTEUNPTU9ETyBDQSBMaW1pdGVkMTgwNgYDVQQD
Ey9DT01PRE8gRUNDIERvbWFpbiBWYWxpZGF0aW9uIFNlY3VyZSBTZXJ2ZXIgQ0Eg
MjAeFw0xODA3MTgwMDAwMDBaFw0xOTAxMjQyMzU5NTlaMGsxITAfBgNVBAsTGERv
bWFpbiBDb250cm9sIFZhbGlkYXRlZDEhMB8GA1UECxMYUG9zaXRpdmVTU0wgTXVs
dGktRG9tYWluMSMwIQYDVQQDExpzbmkyNDE0Mi5jbG91ZGZsYXJlc3NsLmNvbTBZ
MBMGByqGSM49AgEGCCqGSM49AwEHA0IABPt87N1Or6yKVUQkeVvEi7hEpszi3dQG
EGJBKrPGF/7mIXmikqM+8td/Isq8AP9sOiFFjcrexGgTKYNh2em/hGejgge7MIIH
tzAfBgNVHSMEGDAWgBRACWFn8LyDcU/eEggsb9TUK3Y9ljAdBgNVHQ4EFgQUBYbY
tO2pfiPuLud1qjssBggqk7IwDgYDVR0PAQH/BAQDAgeAMAwGA1UdEwEB/wQCMAAw
HQYDVR0lBBYwFAYIKwYBBQUHAwEGCCsGAQUFBwMCME8GA1UdIARIMEYwOgYLKwYB
BAGyMQECAgcwKzApBggrBgEFBQcCARYdaHR0cHM6Ly9zZWN1cmUuY29tb2RvLmNv
bS9DUFMwCAYGZ4EMAQIBMFYGA1UdHwRPME0wS6BJoEeGRWh0dHA6Ly9jcmwuY29t
b2RvY2E0LmNvbS9DT01PRE9FQ0NEb21haW5WYWxpZGF0aW9uU2VjdXJlU2VydmVy
Q0EyLmNybDCBiAYIKwYBBQUHAQEEfDB6MFEGCCsGAQUFBzAChkVodHRwOi8vY3J0
LmNvbW9kb2NhNC5jb20vQ09NT0RPRUNDRG9tYWluVmFsaWRhdGlvblNlY3VyZVNl
cnZlckNBMi5jcnQwJQYIKwYBBQUHMAGGGWh0dHA6Ly9vY3NwLmNvbW9kb2NhNC5j
b20wEwYKKwYBBAHWeQIEAwEB/wQCBQAwggXtBgNVHREEggXkMIIF4IIac25pMjQx
NDIuY2xvdWRmbGFyZXNzbC5jb22CESouYW5pbWVyZWJvcm4uY29tgg4qLmJlZ2xp
ZGVhcy5nYYIMKi5jaHJvbWEuaW5rghQqLmNodWNrc2NsZWFuaW5ncy5nYYIUKi5j
bGlwdnVpZ2lhaXRyaXMuZ2GCDCouY212c2pucy5nYYISKi5jb21wZXRlZ3JhcGhz
LmdhghIqLmNvbnNvbGVwcmludHMuZ2GCECouY29weWJyZWV6ZXMuZ2GCECouY29y
cGhyZXllZHMuZ2GCDiouY3lhbmlnZWVzLmdhgg4qLmRhZHBiZWFycy5nYYIVKi5k
YWh1bGV3b3JsZHdpZGVzLmdhghMqLmRhaWx5b3BlbmluZ3NzLmdhgg4qLmRhbGV5
bGV4cy5nYYISKi5kYW5handlaW5rbGVzLmdhghIqLmRhbmNld3RoeW9nYXMuZ2GC
EiouZGFya21vb3NldnBzcy5nYYIPKi5kYXVyYXQuY29tLmFygg8qLmRlbHRhYmVy
Zy5jb22CEiouZHJqYWhhbm9iZ3lucy5nYYISKi5kcnVua2dpcmxpZXNzLmdhgg4q
LmR1aGllcGt5cy5nYYIPKi5kdWp1YW5qc3FzLmdhghAqLmR1bWJpc2Vhc3lzLmdh
ghQqLmR1bXBzb2Z0ZHJpbmtzcy5nYYITKi5kdW5oYXZlbndvb2Rzcy5nYYIRKi5k
dXJhYmlsaXRlYXMuZ2GCESouZHV4bWFuZ3JvdXBzLmdhghIqLmR2cGRyaXZld2F5
c3MuZ2GCDyouZHdlbGx3aXplcy5nYYINKi5kd3drb3Vpcy5nYYIRKi5lbnRlcnRh
c3RpYy5jb22CFiouZXN0dWRpb2dvbGJlci5jb20uYXKCECoubGV0c3JldHJvLnRl
YW2CESoubWFjY3Vpc2gub3JnLnVrghQqLm1hZGFtc3F1aWdnbGVzLmNvbYIMKi5z
ZnR3Lm5pbmphghAqLnNwYW5nZW5iZXJnLmlvghIqLnRpbW11dHRvbi5jb20uYXWC
FCoud3lvbWluZ3NleGJvb2suY29tggkqLnljaC5iaWSCD2FuaW1lcmVib3JuLmNv
bYIMYmVnbGlkZWFzLmdhggpjaHJvbWEuaW5rghJjaHVja3NjbGVhbmluZ3MuZ2GC
EmNsaXB2dWlnaWFpdHJpcy5nYYIKY212c2pucy5nYYIQY29tcGV0ZWdyYXBocy5n
YYIQY29uc29sZXByaW50cy5nYYIOY29weWJyZWV6ZXMuZ2GCDmNvcnBocmV5ZWRz
LmdhggxjeWFuaWdlZXMuZ2GCDGRhZHBiZWFycy5nYYITZGFodWxld29ybGR3aWRl
cy5nYYIRZGFpbHlvcGVuaW5nc3MuZ2GCDGRhbGV5bGV4cy5nYYIQZGFuYWp3ZWlu
a2xlcy5nYYIQZGFuY2V3dGh5b2dhcy5nYYIQZGFya21vb3NldnBzcy5nYYINZGF1
cmF0LmNvbS5hcoINZGVsdGFiZXJnLmNvbYIQZHJqYWhhbm9iZ3lucy5nYYIQZHJ1
bmtnaXJsaWVzcy5nYYIMZHVoaWVwa3lzLmdhgg1kdWp1YW5qc3FzLmdhgg5kdW1i
aXNlYXN5cy5nYYISZHVtcHNvZnRkcmlua3NzLmdhghFkdW5oYXZlbndvb2Rzcy5n
YYIPZHVyYWJpbGl0ZWFzLmdhgg9kdXhtYW5ncm91cHMuZ2GCEGR2cGRyaXZld2F5
c3MuZ2GCDWR3ZWxsd2l6ZXMuZ2GCC2R3d2tvdWlzLmdhgg9lbnRlcnRhc3RpYy5j
b22CFGVzdHVkaW9nb2xiZXIuY29tLmFygg5sZXRzcmV0cm8udGVhbYIPbWFjY3Vp
c2gub3JnLnVrghJtYWRhbXNxdWlnZ2xlcy5jb22CCnNmdHcubmluamGCDnNwYW5n
ZW5iZXJnLmlvghB0aW1tdXR0b24uY29tLmF1ghJ3eW9taW5nc2V4Ym9vay5jb22C
B3ljaC5iaWQwCgYIKoZIzj0EAwIDSAAwRQIhALh7n3d6ZMt0iDdvwC1mlIilvxiY
1Nvwk/+2PlNhEwQFAiB9pYSkBYOp1S938apWp3n3m5cYNS2iTnioF9+v4dTzpw==
-----END CERTIFICATE-----
```

**Comment by mathiasertl** [comment]
Note that this issue is not yet quite fixed, but will be fixed in the 1.9.0 release. The certificates you posted expose some other (actually unrelated) minor edge cases as well.
